File: mobile/src/components/social/FriendButton.js
import { useState } from 'react';
import { View } from 'react-native';
import Button from '../common/Button';
import * as socialApi from '../../api/social.api';

export default function FriendButton({ profile, onChange }) {
  const [loading, setLoading] = useState(false);

  const run = async (fn) => {
    setLoading(true);
    try {
      await fn();
      onChange?.();
    } finally {
      setLoading(false);
    }
  };

  if (profile.isFriend) {
    return <Button title="Friends ✓" variant="outline" disabled />;
  }

  if (profile.hasPendingFriendRequest?.direction === 'outgoing') {
    return <Button title="Request sent" variant="outline" disabled />;
  }

  if (profile.hasPendingFriendRequest?.direction === 'incoming') {
    const requestId = profile.hasPendingFriendRequest.id;
    return (
      <View style={{ flexDirection: 'row', gap: 8 }}>
        <Button title="Accept" onPress={() => run(() => socialApi.acceptFriendRequest(requestId))} loading={loading} />
        <Button title="Decline" variant="outline" onPress={() => run(() => socialApi.declineFriendRequest(requestId))} loading={loading} />
      </View>
    );
  }

  return <Button title="Add Friend" onPress={() => run(() => socialApi.sendFriendRequest(profile.id))} loading={loading} />;
}

File: mobile/lib/features/profile/widgets/friend_button.dart
import 'package:flutter/material.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';

import '../../../common/widgets/app_button.dart';
import '../../../models/profile.dart';
import '../data/social_api.dart';

/// Ports `src/components/social/FriendButton.js`.
class FriendButton extends ConsumerStatefulWidget {
  final UserProfile profile;
  final VoidCallback? onChange;

  const FriendButton({super.key, required this.profile, this.onChange});

  @override
  ConsumerState<FriendButton> createState() => _FriendButtonState();
}

class _FriendButtonState extends ConsumerState<FriendButton> {
  bool _loading = false;

  Future<void> _run(Future<void> Function() action) async {
    setState(() => _loading = true);
    try {
      await action();
      widget.onChange?.call();
    } finally {
      if (mounted) setState(() => _loading = false);
    }
  }

  @override
  Widget build(BuildContext context) {
    final profile = widget.profile;
    final api = ref.read(socialApiProvider);

    if (profile.isFriend) {
      return const AppButton(label: 'Friends ✓', onPressed: null, variant: AppButtonVariant.outline);
    }

    final pending = profile.hasPendingFriendRequest;
    if (pending?.direction == 'outgoing') {
      return const AppButton(label: 'Request sent', onPressed: null, variant: AppButtonVariant.outline);
    }

    if (pending?.direction == 'incoming') {
      return Row(
        mainAxisSize: MainAxisSize.min,
        children: [
          Expanded(
            child: AppButton(
              label: 'Accept',
              loading: _loading,
              onPressed: _loading ? null : () => _run(() => api.acceptFriendRequest(pending!.id)),
            ),
          ),
          const SizedBox(width: 8),
          Expanded(
            child: AppButton(
              label: 'Decline',
              variant: AppButtonVariant.outline,
              onPressed: _loading ? null : () => _run(() => api.declineFriendRequest(pending!.id)),
            ),
          ),
        ],
      );
    }

    return AppButton(
      label: 'Add Friend',
      loading: _loading,
      onPressed: _loading ? null : () => _run(() => api.sendFriendRequest(profile.id)),
    );
  }
}
